File: src/stream.rs
use async_stream::stream;
use axum::response::{
    sse::{Event, KeepAlive},
    Sse,
};
use eventsource_stream::Eventsource as EventsourceExt;
use futures_util::StreamExt;
use serde_json::{json, Value};
use std::collections::BTreeMap;
use std::sync::Arc;
use tracing::{error, warn};

use crate::{
    session::SessionStore,
    types::{ChatMessage, ChatRequest, ChatStreamChunk},
};

pub struct StreamArgs {
    pub client: reqwest::Client,
    pub url: String,
    pub api_key: Arc<String>,
    pub chat_req: ChatRequest,
    pub response_id: String,
    pub sessions: SessionStore,
    pub prior_messages: Vec<ChatMessage>,
    pub model: String,
}

struct ToolCallAccum {
    id: String,
    name: String,
    arguments: String,
}

/// Translate an upstream Chat Completions SSE stream into a Responses API SSE stream.
///
/// Text response event sequence:
///   response.created → response.output_item.added (message) → response.output_text.delta*
///   → response.output_item.done → response.completed
///
/// Tool call response event sequence:
///   response.created → [accumulate deltas] → response.output_item.added (function_call)
///   → response.function_call_arguments.delta → response.output_item.done → response.completed
pub fn translate_stream(
    args: StreamArgs,
) -> Sse<impl futures_util::Stream<Item = Result<Event, std::convert::Infallible>>> {
    let StreamArgs {
        client,
        url,
        api_key,
        chat_req,
        response_id,
        sessions,
        prior_messages,
        model,
    } = args;
    let msg_item_id = format!("msg_{}", uuid::Uuid::new_v4().simple());

    let event_stream = stream! {
        yield Ok(Event::default()
            .event("response.created")
            .data(json!({
                "type": "response.created",
                "response": { "id": &response_id, "status": "in_progress", "model": &model }
            }).to_string()));

        let mut builder = client.post(&url).header("Content-Type", "application/json");
        if !api_key.is_empty() {
            builder = builder.bearer_auth(api_key.as_str());
        }

        let upstream = match builder.json(&chat_req).send().await {
            Ok(r) if r.status().is_success() => r,
            Ok(r) => {
                let status = r.status();
                let body = r.text().await.unwrap_or_default();
                error!("upstream {status}: {body}");
                yield Ok(Event::default().event("response.failed").data(
                    json!({"type": "response.failed", "response": {"id": &response_id, "status": "failed", "error": {"code": status.as_u16().to_string(), "message": body}}}).to_string()
                ));
                return;
            }
            Err(e) => {
                error!("upstream request failed: {e}");
                yield Ok(Event::default().event("response.failed").data(
                    json!({"type": "response.failed", "response": {"id": &response_id, "status": "failed", "error": {"code": "connection_error", "message": e.to_string()}}}).to_string()
                ));
                return;
            }
        };

        let mut accumulated_text = String::new();
        let mut accumulated_reasoning = String::new();
        let mut tool_calls: BTreeMap<usize, ToolCallAccum> = BTreeMap::new();
        let mut emitted_message_item = false;
        let mut source = upstream.bytes_stream().eventsource();

        while let Some(ev) = source.next().await {
            match ev {
                Err(e) => {
                    warn!("SSE parse error: {e}");
                    break;
                }
                Ok(ev) if ev.data.trim() == "[DONE]" => break,
                Ok(ev) if ev.data.is_empty() => continue,
                Ok(ev) => {
                    match serde_json::from_str::<ChatStreamChunk>(&ev.data) {
                        Err(e) => warn!("chunk parse error: {e} — data: {}", ev.data),
                        Ok(chunk) => {
                            for choice in &chunk.choices {
                                // Reasoning/thinking content (kimi-k2.6 etc.)
                                if let Some(rc) = choice.delta.reasoning_content.as_deref() {
                                    if !rc.is_empty() {
                                        accumulated_reasoning.push_str(rc);
                                    }
                                }

                                // Text content
                                let content = choice.delta.content.as_deref().unwrap_or("");
                                if !content.is_empty() {
                                    if !emitted_message_item {
                                        yield Ok(Event::default()
                                            .event("response.output_item.added")
                                            .data(json!({
                                                "type": "response.output_item.added",
                                                "output_index": 0,
                                                "item": { "type": "message", "id": &msg_item_id, "role": "assistant", "content": [], "status": "in_progress" }
                                            }).to_string()));
                                        emitted_message_item = true;
                                    }
                                    accumulated_text.push_str(content);
                                    yield Ok(Event::default()
                                        .event("response.output_text.delta")
                                        .data(json!({
                                            "type": "response.output_text.delta",
                                            "item_id": &msg_item_id,
                                            "output_index": 0,
                                            "content_index": 0,
                                            "delta": content
                                        }).to_string()));
                                }

                                // Tool call deltas — accumulate by index
                                if let Some(delta_calls) = &choice.delta.tool_calls {
                                    for dc in delta_calls {
                                        let entry = tool_calls.entry(dc.index).or_insert(ToolCallAccum {
                                            id: String::new(),
                                            name: String::new(),
                                            arguments: String::new(),
                                        });
                                        if let Some(id) = &dc.id {
                                            if !id.is_empty() { entry.id.clone_from(id); }
                                        }
                                        if let Some(func) = &dc.function {
                                            if let Some(n) = &func.name {
                                                if !n.is_empty() { entry.name.push_str(n); }
                                            }
                                            if let Some(a) = &func.arguments {
                                                entry.arguments.push_str(a);
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        // Close message item if one was opened
        if emitted_message_item {
            yield Ok(Event::default()
                .event("response.output_item.done")
                .data(json!({
                    "type": "response.output_item.done",
                    "output_index": 0,
                    "item": {
                        "type": "message",
                        "id": &msg_item_id,
                        "role": "assistant",
                        "status": "completed",
                        "content": [{"type": "output_text", "text": &accumulated_text}]
                    }
                }).to_string()));
        }

        // Emit function_call items for each accumulated tool call
        let base_index: usize = if emitted_message_item { 1 } else { 0 };
        let mut fc_items: Vec<Value> = Vec::new();

        for (rel_idx, (_, tc)) in tool_calls.iter().enumerate() {
            let fc_item_id = format!("fc_{}", uuid::Uuid::new_v4().simple());
            let output_index = base_index + rel_idx;

            yield Ok(Event::default()
                .event("response.output_item.added")
                .data(json!({
                    "type": "response.output_item.added",
                    "output_index": output_index,
                    "item": {
                        "type": "function_call",
                        "id": &fc_item_id,
                        "call_id": &tc.id,
                        "name": &tc.name,
                        "arguments": "",
                        "status": "in_progress"
                    }
                }).to_string()));

            if !tc.arguments.is_empty() {
                yield Ok(Event::default()
                    .event("response.function_call_arguments.delta")
                    .data(json!({
                        "type": "response.function_call_arguments.delta",
                        "item_id": &fc_item_id,
                        "output_index": output_index,
                        "delta": &tc.arguments
                    }).to_string()));
            }

            yield Ok(Event::default()
                .event("response.output_item.done")
                .data(json!({
                    "type": "response.output_item.done",
                    "output_index": output_index,
                    "item": {
                        "type": "function_call",
                        "id": &fc_item_id,
                        "call_id": &tc.id,
                        "name": &tc.name,
                        "arguments": &tc.arguments,
                        "status": "completed"
                    }
                }).to_string()));

            fc_items.push(json!({
                "type": "function_call",
                "id": fc_item_id,
                "call_id": &tc.id,
                "name": &tc.name,
                "arguments": &tc.arguments,
                "status": "completed"
            }));
        }

        // Persist turn to session store
        // Store reasoning_content per call_id so translate.rs can inject it
        // back when Codex replays function_call items in the next request.
        for tc in tool_calls.values() {
            if !tc.id.is_empty() {
                sessions.store_reasoning(tc.id.clone(), accumulated_reasoning.clone());
            }
        }

        let mut messages = prior_messages;
        let assistant_tool_calls: Option<Vec<Value>> = if tool_calls.is_empty() {
            None
        } else {
            Some(tool_calls.values().map(|tc| json!({
                "id": &tc.id,
                "type": "function",
                "function": { "name": &tc.name, "arguments": &tc.arguments }
            })).collect())
        };
        messages.push(ChatMessage {
            role: "assistant".into(),
            content: if accumulated_text.is_empty() { None } else { Some(accumulated_text.clone()) },
            reasoning_content: if accumulated_reasoning.is_empty() { None } else { Some(accumulated_reasoning.clone()) },
            tool_calls: assistant_tool_calls,
            tool_call_id: None,
            name: None,
        });
        sessions.save_with_id(response_id.clone(), messages);

        // Build output array for response.completed
        let mut output_items: Vec<Value> = Vec::new();
        if emitted_message_item {
            output_items.push(json!({
                "type": "message",
                "id": &msg_item_id,
                "role": "assistant",
                "status": "completed",
                "content": [{"type": "output_text", "text": &accumulated_text}]
            }));
        }
        output_items.extend(fc_items);

        yield Ok(Event::default()
            .event("response.completed")
            .data(json!({
                "type": "response.completed",
                "response": {
                    "id": &response_id,
                    "status": "completed",
                    "model": &model,
                    "output": output_items
                }
            }).to_string()));
    };

    Sse::new(event_stream).keep_alive(KeepAlive::default())
}

File: stream.ts
import type { ChatRequest, ChatMessage, ChatStreamChunk, DeltaToolCall, ChatLogEntry } from "./types.ts";
import type { UserConfig } from "./config.ts";
import { Config, SessionStore } from "./config.ts";
import { createModuleLogger } from "./logger.ts";
import { UsageStore } from "./usage.ts";
import { appendChatLog, appendSystemError } from "./persist.ts";

const log = createModuleLogger("stream");

export interface StreamArgs {
  url: string;
  apiKey: string;
  fallbackApiKey?: string;
  chatReq: ChatRequest;
  responseId: string;
  sessions: SessionStore;
  priorMessages: ChatMessage[];
  requestMessages: ChatMessage[];
  codexModel: string;
  upstreamModel: string;
  user: UserConfig;
  config: Config;
  usageStore: UsageStore;
}

interface ToolCallAccum {
  id: string;
  name: string;
  arguments: string;
}

/// Translate an upstream Chat Completions SSE stream into a Responses API SSE stream.
/// Text response event sequence:
///   response.created → response.output_item.added (message) → response.output_text.delta*
///   → response.output_item.done → response.completed
///
/// Tool call response event sequence:
///   response.created → [accumulate deltas] → response.output_item.added (function_call)
///   → response.function_call_arguments.delta → response.output_item.done → response.completed
export async function* translateStream(
  args: StreamArgs
): AsyncGenerator<string, void, unknown> {
  const {
    url,
    apiKey,
    fallbackApiKey,
    chatReq,
    responseId,
    sessions,
    priorMessages,
    requestMessages,
    codexModel,
    upstreamModel,
    user,
    config,
    usageStore,
  } = args;

  const msgItemId = `msg_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;

  // Emit response.created
  yield sseEvent("response.created", {
    type: "response.created",
    response: { id: responseId, status: "in_progress", model: codexModel },
  });

  log.info("开始流式请求", { responseId, model: codexModel, url, user: user.name });

  // Define interruptive error codes that trigger fallback
  const INTERRUPTIVE_STATUS_CODES = [502, 503, 504, 429];

  let linkType: "primary" | "fallback" = "primary";
  let fallbackReason: string | undefined;
  let usedApiKey = apiKey;

  // Send request to upstream
  const headers: Record<string, string> = { "Content-Type": "application/json" };
  if (usedApiKey) headers["Authorization"] = `Bearer ${usedApiKey}`;

  let upstream: Response;
  let primaryError: { status: number; error: string } | null = null;

  // Primary attempt
  try {
    upstream = await fetch(url, {
      method: "POST",
      headers,
      body: JSON.stringify(chatReq),
    });

    if (!upstream.ok) {
      primaryError = { status: upstream.status, error: await upstream.text().catch(() => "") };

      // Check if we should trigger fallback
      if (INTERRUPTIVE_STATUS_CODES.includes(upstream.status) && fallbackApiKey) {
        log.warn("流式 Primary 失败，切换 Fallback", {
          status: upstream.status,
          user: user.name,
          model: codexModel,
        });

        // Emit fallback triggered event
        yield sseEvent("response.fallback_triggered", {
          type: "response.fallback_triggered",
          response: { id: responseId },
          reason: `primary_${upstream.status}`,
        });

        // Fallback attempt
        const fallbackHeaders: Record<string, string> = { "Content-Type": "application/json" };
        fallbackHeaders["Authorization"] = `Bearer ${fallbackApiKey}`;

        try {
          upstream = await fetch(url, {
            method: "POST",
            headers: fallbackHeaders,
            body: JSON.stringify(chatReq),
          });

          if (upstream.ok) {
            linkType = "fallback";
            fallbackReason = `primary_${primaryError.status}`;
            usedApiKey = fallbackApiKey;
            log.info("流式 Fallback 成功", { user: user.name, model: codexModel });
          } else {
            // Fallback also failed
            const fallbackError = await upstream.text().catch(() => "");
            log.error("流式 Fallback 也失败", { status: upstream.status, user: user.name });

            // Record system error
            await appendSystemError(config.dataDir, {
              timestamp: new Date().toISOString(),
              type: "upstream_error",
              error: `primary_${primaryError.status}, fallback_${upstream.status}`,
              upstream: config.upstream,
              model: upstreamModel,
              request_id: responseId,
              user: user.name,
            });

            yield sseEvent("response.failed", {
              type: "response.failed",
              response: {
                id: responseId,
                status: "failed",
                error: { code: String(upstream.status), message: fallbackError },
              },
            });
            return;
          }
        } catch (fallbackErr) {
          log.error("流式 Fallback 连接失败", { error: String(fallbackErr), user: user.name });

          await appendSystemError(config.dataDir, {
            timestamp: new Date().toISOString(),
            type: "upstream_error",
            error: `primary_${primaryError.status}, fallback_${String(fallbackErr)}`,
            upstream: config.upstream,
            model: upstreamModel,
            request_id: responseId,
            user: user.name,
          });

          yield sseEvent("response.failed", {
            type: "response.failed",
            response: {
              id: responseId,
              status: "failed",
              error: { code: "fallback_connection_error", message: String(fallbackErr) },
            },
          });
          return;
        }
      } else {
        // No fallback or non-interruptive error
        log.error("上游返回错误", { status: primaryError.status, user: user.name });

        // Record system error for 5xx without fallback
        if (primaryError.status >= 500 && !fallbackApiKey) {
          await appendSystemError(config.dataDir, {
            timestamp: new Date().toISOString(),
            type: "upstream_error",
            error: `${primaryError.status}: ${primaryError.error}`,
            upstream: config.upstream,
            model: upstreamModel,
            request_id: responseId,
            user: user.name,
          });
        }

        yield sseEvent("response.failed", {
          type: "response.failed",
          response: {
            id: responseId,
            status: "failed",
            error: { code: String(primaryError.status), message: primaryError.error },
          },
        });
        return;
      }
    }
  } catch (e) {
    log.error("上游连接失败", { error: String(e), user: user.name });
    primaryError = { status: 0, error: String(e) };

    // Network error - try fallback if available
    if (fallbackApiKey) {
      log.warn("流式 Primary 连接异常，切换 Fallback", { user: user.name, model: codexModel });

      yield sseEvent("response.fallback_triggered", {
        type: "response.fallback_triggered",
        response: { id: responseId },
        reason: "primary_connection_error",
      });

      const fallbackHeaders: Record<string, string> = { "Content-Type": "application/json" };
      fallbackHeaders["Authorization"] = `Bearer ${fallbackApiKey}`;

      try {
        upstream = await fetch(url, {
          method: "POST",
          headers: fallbackHeaders,
          body: JSON.stringify(chatReq),
        });

        if (upstream.ok) {
          linkType = "fallback";
          fallbackReason = "primary_connection_error";
          usedApiKey = fallbackApiKey;
          log.info("流式 Fallback 成功", { user: user.name, model: codexModel });
        } else {
          const fallbackError = await upstream.text().catch(() => "");
          log.error("流式 Fallback 也失败", { status: upstream.status, user: user.name });

          await appendSystemError(config.dataDir, {
            timestamp: new Date().toISOString(),
            type: "upstream_error",
            error: `primary_connection_error, fallback_${upstream.status}`,
            upstream: config.upstream,
            model: upstreamModel,
            request_id: responseId,
            user: user.name,
          });

          yield sseEvent("response.failed", {
            type: "response.failed",
            response: {
              id: responseId,
              status: "failed",
              error: { code: String(upstream.status), message: fallbackError },
            },
          });
          return;
        }
      } catch (fallbackErr) {
        log.error("流式 Fallback 连接也异常", { error: String(fallbackErr), user: user.name });

        await appendSystemError(config.dataDir, {
          timestamp: new Date().toISOString(),
          type: "upstream_error",
          error: `primary_connection_error, fallback_${String(fallbackErr)}`,
          upstream: config.upstream,
          model: upstreamModel,
          request_id: responseId,
          user: user.name,
        });

        yield sseEvent("response.failed", {
          type: "response.failed",
          response: {
            id: responseId,
            status: "failed",
            error: { code: "fallback_connection_error", message: String(fallbackErr) },
          },
        });
        return;
      }
    } else {
      await appendSystemError(config.dataDir, {
        timestamp: new Date().toISOString(),
        type: "upstream_error",
        error: String(e),
        upstream: config.upstream,
        model: upstreamModel,
        request_id: responseId,
        user: user.name,
      });

      yield sseEvent("response.failed", {
        type: "response.failed",
        response: {
          id: responseId,
          status: "failed",
          error: { code: "connection_error", message: String(e) },
        },
      });
      return;
    }
  }

  log.debug("上游连接成功", { status: upstream.status, link_type: linkType, user: user.name });

  const reader = upstream.body?.getReader();
  if (!reader) {
    log.error("无法获取响应 body");
    yield sseEvent("response.failed", {
      type: "response.failed",
      response: {
        id: responseId,
        status: "failed",
        error: { code: "no_body", message: "No response body" },
      },
    });
    return;
  }

  const decoder = new TextDecoder();
  let buffer = "";
  let accumulatedText = "";
  let accumulatedReasoning = "";
  const toolCalls: Map<number, ToolCallAccum> = new Map();
  let emittedMessageItem = false;
  let accumulatedUsage = { prompt_tokens: 0, completion_tokens: 0, total_tokens: 0 };

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });

      // Process SSE events in buffer
      const lines = buffer.split("\n");
      buffer = lines.pop() || ""; // Keep incomplete line in buffer

      for (const line of lines) {
        if (line.startsWith("data: ")) {
          const data = line.slice(6).trim();
          if (data === "[DONE]") continue;
          if (!data) continue;

          try {
            const chunk: ChatStreamChunk = JSON.parse(data);

            // Accumulate usage if present (usually in final chunk)
            if (chunk.usage) {
              accumulatedUsage.prompt_tokens = chunk.usage.prompt_tokens;
              accumulatedUsage.completion_tokens = chunk.usage.completion_tokens;
              accumulatedUsage.total_tokens = chunk.usage.total_tokens;
            }

            for (const choice of chunk.choices || []) {
              // Reasoning/thinking content (kimi-k2.6 etc.)
              const rc = choice.delta?.reasoning_content;
              if (rc) accumulatedReasoning += rc;

              // Text content
              const content = choice.delta?.content || "";
              if (content) {
                if (!emittedMessageItem) {
                  yield sseEvent("response.output_item.added", {
                    type: "response.output_item.added",
                    output_index: 0,
                    item: {
                      type: "message",
                      id: msgItemId,
                      role: "assistant",
                      content: [],
                      status: "in_progress",
                    },
                  });
                  emittedMessageItem = true;
                }
                accumulatedText += content;
                yield sseEvent("response.output_text.delta", {
                  type: "response.output_text.delta",
                  item_id: msgItemId,
                  output_index: 0,
                  content_index: 0,
                  delta: content,
                });
              }

              // Tool call deltas — accumulate by index
              const deltaCalls = choice.delta?.tool_calls;
              if (deltaCalls) {
                for (const dc of deltaCalls) {
                  const entry = toolCalls.get(dc.index) || {
                    id: "",
                    name: "",
                    arguments: "",
                  };
                  if (dc.id) entry.id = dc.id;
                  if (dc.function?.name) entry.name += dc.function.name;
                  if (dc.function?.arguments) entry.arguments += dc.function.arguments;
                  toolCalls.set(dc.index, entry);
                }
              }
            }
          } catch {
            // Skip invalid JSON
          }
        }
      }
    }
  } finally {
    reader.releaseLock();
  }

  // Close message item if one was opened
  if (emittedMessageItem) {
    yield sseEvent("response.output_item.done", {
      type: "response.output_item.done",
      output_index: 0,
      item: {
        type: "message",
        id: msgItemId,
        role: "assistant",
        status: "completed",
        content: [{ type: "output_text", text: accumulatedText }],
      },
    });
  }

  // Emit function_call items for each accumulated tool call
  const baseIndex = emittedMessageItem ? 1 : 0;
  const fcItems: unknown[] = [];

  const toolCallEntries = Array.from(toolCalls.entries());
  for (const [relIdx, [_, tc]] of toolCallEntries.entries()) {
    const fcItemId = `fc_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
    const outputIndex = baseIndex + relIdx;

    yield sseEvent("response.output_item.added", {
      type: "response.output_item.added",
      output_index: outputIndex,
      item: {
        type: "function_call",
        id: fcItemId,
        call_id: tc.id,
        name: tc.name,
        arguments: "",
        status: "in_progress",
      },
    });

    if (tc.arguments) {
      yield sseEvent("response.function_call_arguments.delta", {
        type: "response.function_call_arguments.delta",
        item_id: fcItemId,
        output_index: outputIndex,
        delta: tc.arguments,
      });
    }

    yield sseEvent("response.output_item.done", {
      type: "response.output_item.done",
      output_index: outputIndex,
      item: {
        type: "function_call",
        id: fcItemId,
        call_id: tc.id,
        name: tc.name,
        arguments: tc.arguments,
        status: "completed",
      },
    });

    fcItems.push({
      type: "function_call",
      id: fcItemId,
      call_id: tc.id,
      name: tc.name,
      arguments: tc.arguments,
      status: "completed",
    });
  }

  log.info("流处理完成", {
    responseId,
    text_len: accumulatedText.length,
    reasoning_len: accumulatedReasoning.length,
    tool_calls_count: toolCalls.size,
    usage: accumulatedUsage,
    link_type: linkType,
    user: user.name,
  });

  // Persist turn to session store
  for (const tc of toolCalls.values()) {
    if (tc.id) {
      sessions.storeReasoning(tc.id, accumulatedReasoning);
    }
  }

  const assistantToolCalls = toolCalls.size > 0
    ? Array.from(toolCalls.values()).map((tc) => ({
        id: tc.id,
        type: "function",
        function: { name: tc.name, arguments: tc.arguments },
      }))
    : undefined;

  const assistantMsg: ChatMessage = {
    role: "assistant",
    content: accumulatedText || undefined,
    reasoning_content: accumulatedReasoning || undefined,
    tool_calls: assistantToolCalls,
  };

  // Index reasoning by turn fingerprint
  if (accumulatedReasoning) {
    sessions.storeTurnReasoning(requestMessages, assistantMsg, accumulatedReasoning);
  }

  const messages = [...priorMessages, assistantMsg];
  sessions.saveWithId(responseId, messages);

  log.debug("会话已存储", { responseId, messages_count: messages.length });

  // Record usage
  await usageStore.recordUsage({
    user: user.name,
    timestamp: new Date().toISOString(),
    input_tokens: accumulatedUsage.prompt_tokens,
    output_tokens: accumulatedUsage.completion_tokens,
    total_tokens: accumulatedUsage.total_tokens,
    model: codexModel,
    upstream_model: upstreamModel,
    link_type: linkType,
    request_id: responseId,
    fallback_reason: fallbackReason,
  });

  // Record chat log
  const chatLogEntry: ChatLogEntry = {
    timestamp: new Date().toISOString(),
    user: user.name,
    request_id: responseId,
    model: codexModel,
    upstream_model: upstreamModel,
    link_type: linkType,
    request: { model: chatReq.model, messages: chatReq.messages, stream: true },
    response: {
      text: accumulatedText,
      reasoning: accumulatedReasoning,
      tool_calls: assistantToolCalls,
      usage: accumulatedUsage,
    },
    usage: { input_tokens: accumulatedUsage.prompt_tokens, output_tokens: accumulatedUsage.completion_tokens },
    fallback_reason: fallbackReason,
  };
  await appendChatLog(config.dataDir, user.name, chatLogEntry);

  // Build output array for response.completed
  const outputItems: unknown[] = [];
  if (emittedMessageItem) {
    outputItems.push({
      type: "message",
      id: msgItemId,
      role: "assistant",
      status: "completed",
      content: [{ type: "output_text", text: accumulatedText }],
    });
  }
  outputItems.push(...fcItems);

  yield sseEvent("response.completed", {
    type: "response.completed",
    response: {
      id: responseId,
      status: "completed",
      model: codexModel,
      output: outputItems,
      usage: {
        input_tokens: accumulatedUsage.prompt_tokens,
        output_tokens: accumulatedUsage.completion_tokens,
        total_tokens: accumulatedUsage.total_tokens,
      },
    },
  });
}

/// Format an SSE event string
function sseEvent(event: string, data: unknown): string {
  return `event: ${event}\ndata: ${JSON.stringify(data)}\n\n`;
}